src/box/lua/console.lua

```diff
@@ -818,6 +818,9 @@ local function connect(uri, opts)
     if u == nil or u.service == nil then
         error('Usage: console.connect("[login:password@][host:]port")')
     end
+    if not u.host then
+        u.host = 'localhost'
+    end
```

Comment:
AFAIU, it works till some version, but degraded then. Did you look, what was the reason of the degradation?

Reply:
At the moment, the logic of the work looks like this:

If we call `tarantoolctl connect 3301`, then after [this](https://github.com/tarantool/tarantool/blob/master/src/box/lua/console.lua#L816) we have `u = {["service"] = 3301}`.

Next , the sequence of function calls with arguments is as follows:

* net_box.connect(u.host=`nil`, u.service=`3301`, {})
* net_box.parse_connect_params(host_or_uri=`nil`, ...)

According to the logic of net_box.parse_connect_params, `host_or_uri` gets from `u.host`.

I believe that my solution that has been proposed with u.host=`localhost` before `net_box.connect` call allows to correct the situation. This default assignment goes next to the call of `urilib.parse(tostring(uri))` and checking whether the user forgot to specify the port. The location in this part of the code is expected from the point of view of code readability and [this](https://github.com/tarantool/tarantool/blob/master/src/box/lua/console.lua#L819) requirement.

Now the rationale for why  it worked before and stopped now:

AFAIU, the sequence of function calls with arguments after `urilib.parse` was as follows (everything is also being considered `tarantoolctl connect 3301`):

* net_box.establish_connection(u.host=`nil`, u.service=`3301`, opts.timeout=...)
* socket.tcp_connect(host=u.host=`nil`, post=u.service=`3301`, ...)
* socket.getaddrinfo(host=`nil`, port=`3301`, ...)

So, [socket.getaddrinfo](https://www.tarantool.io/en/doc/latest/reference/reference_lua/socket/#socket-getaddrinfo) return info about `localhost` when we call it with host=`nil`. 

Now we crash at [this](https://github.com/tarantool/tarantool/blob/master/src/box/lua/net_box.lua#L145) moment, and up to this moment the function `socket.getaddrinfo` is not called, so this bug has appeared.

Reply:
I was interested in a commit that introduces the regression. My testing code and `git bisect` results are below.

Run tarantool that listens on 3301 (I have `box.schema.user.grant('guest', 'super')` written to the snapshot):

```sh
$ tarantool -e 'box.cfg({listen = 3301})'
```

Run the test:

```sh
$ ./src/tarantool extra/dist/tarantoolctl connect 3301
```

| Commit                      | Result |
| --------------------------- | ------ |
| 2.10.0-beta2-0-g9d80b68b9   | good   |
| 2.10.0-beta2-1-g0badb4130   | bad    |
| 2.10.0-beta2-2-g8b1aef56a   | bad    |
| 2.10.0-beta2-5-gdc19be406   | bad    |
| 2.10.0-beta2-11-ge493b777d  | bad    |
| 2.10.0-beta2-23-gc9d186082  | bad    |
| 2.10.0-beta2-46-gefee67ca9  | bad    |
| 2.10.0-beta2-92-gee6de025f  | bad    |
| 2.10.0-beta2-184-g2c0aedca0 | bad    |
| 2.10.0-rc1-0-g7ed15e62b     | bad    |

So the problem was introduced by this commit: 0badb4130deac46d6651cdcecb6c6d7ee401a96a. So net.box itself is likely affected and maybe it worth to return the old behaviour back in it. Anyway, I didn't dig into details much and I know who knows all the details: the author of the commit. Please, ask a review from him.

Reply:
AFAIK `localhost` isn't quite the same as no host, see [`getaddrinfo(3)`](https://man7.org/linux/man-pages/man3/getaddrinfo.3.html). I think we should fix the net.box to accept nil for host:

```diff
diff --git a/src/box/lua/net_box.lua b/src/box/lua/net_box.lua
index 570108fa5380..b9c4fc2c9324 100644
--- a/src/box/lua/net_box.lua
+++ b/src/box/lua/net_box.lua
@@ -138,8 +138,8 @@ local function parse_connect_params(host_or_uri, ...) -- self? host_or_uri port?
                         type(host_or_uri) == 'number' or
                         type(host_or_uri) == 'table') then
         uri = host_or_uri
-    elseif type(host_or_uri) == 'string' and (type(port) == 'string' or
-                                              type(port) == 'number') then
+    elseif (type(host_or_uri) == 'string' or host_or_uri == nil) and
+           (type(port) == 'string' or type(port) == 'number') then
         uri = urilib.format({host = host_or_uri, service = tostring(port)})
     else
         box.error(E_PROC_LUA,
```

This won't work though, because surprisingly `uri.format` doesn't work if there's no host given, just a service:

```
tarantool> uri.parse(uri.format(uri.parse(3301)))
---
- null
- 'Incorrect URI: expected host:service or /unix.socket'
...
```

This should probably be fixed, as well.